### PR TITLE
Spotlight null query case

### PR DESCRIPTION
**Pull request by nickoferrall**
WIP for issue https://github.com/ParabolInc/parabol/issues/5183

### CODE REVIEW

File: packages/client/types/constEnums.ts [hunk 1]
```diff
@@ -391,9 +391,9 @@ export const enum ZIndex {
   SIDEBAR = 16,
   SIDE_SHEET = 8,
   SNACKBAR = 24 /* snackbar is kind of dialog */,
-  SPOTLIGHT_MODAL = 25, // above the tooltip
   REFLECTION_IN_FLIGHT = 8,
-  REFLECTION_IN_FLIGHT_LOCAL = 26, // keep it above the dialog
+  REFLECTION_IN_FLIGHT_LOCAL = 28, // keep it above the dialog
+  REFLECTION_IN_FLIGHT_SPOTLIGHT = 26,
```
Comment (nickoferrall) [inline packages/client/types/constEnums.ts:396]: Remote reflections that are in the Spotlight (but not those that aren't in the Spotlight) need to appear above the Spotlight modal but below local reflections. The Spotlight modal uses usePortal which uses the dialog zIndex of 24. usePortal is also used by expandedReflectionsStack and a bunch of other modals which, unlike the Spotlight, need to appear above the remote reflections.

I think the cleanest approach is adding a `REFLECTION_IN_FLIGHT_SPOTLIGHT` zIndex and then passing an `isInSpotlight` prop into `useDraggableReflectionCard`

File: packages/client/mutations/EndDraggingReflectionMutation.ts [hunk 2]
```diff
@@ -87,6 +87,42 @@ const handleAddReflectionGroupToGroups = (store, reflectionGroup) => {
...
+    searchQuery: ''
+  })
+  const wasInSpotlightGroups = similarReflectionGroups?.find(
+    (group) => group.getValue('id') === oldReflectionGroupId
+  )
+  const isInSpotlightGroups = similarReflectionGroups?.find(
+    (group) => group.getValue('id') === reflectionGroupId
+  )
+  // check if ungrouping is happening in the spotlight
+  if (!isInSpotlightGroups && wasInSpotlightGroups) {
```
Comment (nickoferrall) [inline packages/client/mutations/EndDraggingReflectionMutation.ts:116]: If the following scenario happens:
- the Spotlight opens and has 10 reflection groups
- a user on another browser adds a reflection called "dogs" into one of the 10 Spotlight groups
- that user then removes the reflection "dogs" from that group

Would you expect the "dogs" reflection to remain in the Spotlight so that there are 11 groups or that it returns to the kanban? I think it should probably stay (which is how the current implementation works) because while it could have been a accidentally grouped, it's probably more likely that it's a relevant reflection group

File: packages/client/components/ReflectionGroup/DraggableReflectionCard.tsx [hunk 3]
```diff
@@ -89,15 +92,17 @@ const DraggableReflectionCard = (props: Props) => {
...
   )
+
   const isDragPhase = phaseType === 'group' && !isComplete
   const canDrag = isDraggable && isDragPhase && !isEditing && !isDropping
   // slow state updates can mean we miss an onMouseDown event, so use isDragPhase instead of canDrag
   const handleDrag = isDragPhase ? onMouseDown : undefined
   return (
     <DragWrapper
-      ref={(c) => (drag.ref = c)}
+      ref={(c) => (drag.ref = isBehindSpotlight ? null : c)}
```
Comment (nickoferrall) [inline packages/client/components/ReflectionGroup/DraggableReflectionCard.tsx:105]: This fixed a tricky bug. When dragging a reflection group in the Spotlight, `useDraggableReflectionCards` would use the ref that's in the kanban instead of the Spotlight. This made dragging, grouping, and ungrouping very buggy. 

Remote reflections still begin from the kanban position which I think is preferable as it indicates that that's where the teammate is dragging the card from

File: packages/server/graphql/types/User.ts [hunk 4]
```diff
@@ -421,19 +423,38 @@ const User = new GraphQLObjectType<any, GQLContext>({
...
+            .run()
         ])
         const {meetingId: viewerMeetingId} = viewerMeetingMember
         if (viewerMeetingId !== meetingId) {
-          standardError(new Error('Not on team'), {userId})
-          return []
+          return standardError(new Error('Not on team'), {userId})
         }
-        return reflectionGroups
+        const {groups} = groupReflections(reflections, AUTO_GROUPING_THRESHOLD)
```
Comment (nickoferrall) [inline packages/server/graphql/types/User.ts:445]: Hey @tianrunhe, it looks like you've played with the `groupReflection` function before. I'd love to get your help here.

The AC says:
> The heuristic above should attempt to return 10 reflections & if the agglomeration returns less (and there are more than 10 other reflection groups in the meeting), then increase the grouping threshold iteratively until it works

However, if I change the grouping threshold from 0.25 to 0.95 or 0.05, I get no difference in results. The same number of reflections and reflection groups are returned. See [the issue here](https://github.com/ParabolInc/parabol/issues/5183) for more context and screenshots. Perhaps I'm missing something?

Comment (mattkrick) [reply packages/server/graphql/types/User.ts:445]: what text are you using for the reflections?
the text has to be long enough & meaningful enough to actually work. it's gonna use a string similarity algo to detect how similar things are. if the reflection texts are too small, the similarity scores will be 0. might be worth using data from real retros to play with

File: packages/client/mutations/EndDraggingReflectionMutation.ts [hunk 5]
```diff
@@ -87,6 +87,42 @@ const handleAddReflectionGroupToGroups = (store, reflectionGroup) => {
   addNodeToArray(reflectionGroup, meeting, 'reflectionGroups', 'sortOrder')
 }
 
+const handleAddReflectionToSpotlight = (
```
Comment (nickoferrall) [inline packages/client/mutations/EndDraggingReflectionMutation.ts:90]: Hey @mattkrick, I've got two architectural questions I'd like your advice on:

1. Should the remote reflection position begin where the remote user is dragging the reflection, i.e. the kanban, or in the Spotlight? 

Currently, it begins in the kanban (see gif). This tells the user it's being dragged by a teammate but maybe the position feels a bit random.

2. Should the results be groupable/ungroupable?
Enrique mentioned [here](https://github.com/ParabolInc/parabol/issues/4952#issuecomment-862513558) that we should prevent users from grouping/ungrouping the results.
This causes a problem in the following scenario:
- I have the Spotlight open
- another user ungroups a group on their browser that's in my Spotlight results
- that creates a new reflection group id which does not exist in the similarReflectionGroups query
- in my Spotlight, it looks like the card disappears

In this PR, if a results group is ungrouped, I'm adding the newly created reflection group id to the similarReflectionGroups query. This way, the card doesn't disappear and instead remains on the Spotlight.

Alternatively, if the remote reflection begins in the Spotlight, perhaps it could be dragged to the kanban and then disappear from the Spotlight. To do this, I'd try using the ref of the Spotlight card [here](https://github.com/ParabolInc/parabol/blob/0ef8108f6fc8cfce1c59d2fcd40d65a362acee4b/packages/client/components/ReflectionGroup/RemoteReflection.tsx#L63) instead of the `remoteDrag` position.

remote-reflections

Note: I'll create a separate PR with a better Spotlight groups UI

File: packages/server/package.json [hunk 6]
```diff
@@ -131,9 +131,9 @@
     "relay-runtime": "^11.0.2",
     "rethinkdb-ts": "2.4.5",
     "rethinkdbdash": "^2.3.31",
-    "samlify": "^2.6.1",
+    "samlify": "2.7.7",
     "secure-compare": "^3.0.1",
-    "sharp": "^0.26.2",
+    "sharp": "^0.28.0",
```
Comment (nickoferrall) [inline packages/server/package.json:136]: Needed for the m1 chip. I'll create a separate PR with these changes once this is merged: https://github.com/ParabolInc/parabol/pull/5396

File: packages/client/components/ReflectionCard/ReflectionCard.tsx [hunk 7]
```diff
@@ -230,12 +224,13 @@ const ReflectionCard = (props: Props) => {
...
 
-  const showSpotlight = false
+  const showSpotlight = !__PRODUCTION__
   const showSearch =
     phaseType === 'group' &&
-    !inSpotlight &&
+    !isSpotlightOpen &&
     !isComplete &&
     showSpotlight &&
+    !isDemoRoute() &&
```
Comment (nickoferrall) [inline packages/client/components/ReflectionCard/ReflectionCard.tsx:233]: The demo only has a small number of reflections so the Spotlight isn't necessary although it might still be nice to show it off. If you think it's worth including, I suggest that I create a new issue for the demo and tackle that as the last issue in the Spotlight

File: packages/client/mutations/EndDraggingReflectionMutation.ts (hunk 5 above)
Comment (nickoferrall) [reply packages/client/mutations/EndDraggingReflectionMutation.ts:90]: I think the answers to both 1 & 2 are yes. I've pushed an update following the remote reflection plan I outlined here. In the Spotlight UI groups PR, I'll work on adding new reflection groups to the Spotlight column with the least groups

**Comment by nickoferrall** [comment]
Hey @tianrunhe, would you mind reviewing this PR, please? Specifically, it'd be great to get your eyes on the `similarReflectionsQuery` logic in the `User` query and to test that the AC in the issue has been completed

**Review by tianrunhe — CHANGES_REQUESTED** [review]
- [x] if the searchQuery is an empty string, then we use hierarchical agglomeration to filter & sort the reflectionGroups in the meeting (see https://github.com/ParabolInc/parabol/blob/db45e3b5a4c27e917dcc267d33175bae8a04970b/packages/server/graphql/mutations/helpers/handleCompletedStage.ts#L40 as an example)
- [x] The heuristic above should attempt to return 10 reflections & if the agglomeration returns less (and there are more than 10 other reflection groups in the meeting), then increase the grouping threshold iteratively until it works
- [x] on the front end, the groups show up as expected in the results space
- [x] on another browser, if 2 of those groups are merged in real time, that change is reflected in the search results
- [x] on another browser, if 1 of those reflections is ungrouped, then that change is reflected in the search results

File: packages/server/graphql/types/User.ts [hunk 8]
```diff
@@ -421,19 +423,57 @@ const User = new GraphQLObjectType<any, GQLContext>({
...
-          dataLoader.get('retroReflectionGroupsByMeetingId').load(meetingId)
+          r
+            .table('RetroReflection')
+            .getAll(meetingId, {index: 'meetingId'})
+            .filter({isActive: true})
+            .orderBy('createdAt')
+            .run()
         ])
         const {meetingId: viewerMeetingId} = viewerMeetingMember
         if (viewerMeetingId !== meetingId) {
```
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/types/User.ts:443]: The `meetingId` should always be the same but the `viewerMeetingMember` uses the userId that we get from the source. If the viewer passes a `reflectionId` to the query that lives in a meeting that they're not a part of, the viewer's meeting id should be different.

Now that I'm looking at it, I don't like using `toTeamMemberId` for meeting members so I've added a new helper function

File: packages/server/graphql/types/User.ts [hunk 9]
```diff
@@ -421,19 +423,57 @@ const User = new GraphQLObjectType<any, GQLContext>({
...
+        while (currentThreshold) {
+          const {groupedReflections, nextThresh} = groupReflections(reflections, {
+            groupingThreshold: currentThreshold,
+            maxGroupSize: reflectionsCount,
+            maxReductionPercent: 1
+          })
+          const spotlightGroup = groupedReflections.find(
+            (group) => group.reflectionId === reflectionId
+          )
+          if (!spotlightGroup) break
```
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/types/User.ts:461]: It should always be non-null but TypeScript doesn't know that, unfortunately. I could either use a bang after the `spotlightGroup` or break here

File: packages/server/graphql/types/User.ts [hunk 10]
```diff
@@ -421,19 +423,57 @@ const User = new GraphQLObjectType<any, GQLContext>({
...
+            }
+            if (
+              group.reflectionGroupId === spotlightGroup.reflectionGroupId &&
+              group.oldReflectionGroupId !== spotlightGroup.oldReflectionGroupId
+            ) {
+              similarGroupIds.add(group.oldReflectionGroupId)
+            }
+            currentThreshold = nextThresh
+          }
+        }
```
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/types/User.ts]: In the while loop, we're looking through all of the `groupedReflections` rather than the `reflectionGroups`. `spotlightGroup` is the grouped reflection in the source (top bit) of the Spotlight. If we break the loop when we reach 10 `groupedReflections`, there may be less than 10 `reflectionGroups` as several reflections could be in one group. I think we need to keep track of the unique old `reflectionGroupIds` in the loop so that we can guarantee 10 groups.

I agree it's confusing - I've pushed an update changing group to `groupedReflection` to try to add more clarity. Let me know what ya think!

**Comment by nickoferrall** [comment]
Hey @tianrunhe, thanks for the review! I've pushed an update and have responded to the comments. I know you're OOO for a couple of days so if you don't get the chance to review, please let me know and I'll request a different reviewer to keep things moving

**Comment by tianrunhe** [comment]
> Hey @tianrunhe, thanks for the review! I've pushed an update and have responded to the comments. I know you're OOO for a couple of days so if you don't get the chance to review, please let me know and I'll request a different reviewer to keep things moving 

Yeah sorry man, I don't wanna hold you back! I think it's time for you to make some new friends 🤣. I'll definitely check it again when I'm back next Wednesday!

File: packages/client/mutations/EndDraggingReflectionMutation.ts (hunk 5 above)
Comment (mattkrick) [reply packages/client/mutations/EndDraggingReflectionMutation.ts:90]: > Should the remote reflection position begin where the remote user is dragging the reflection, i.e. the kanban, or in the Spotlight?

The trick here is a single source of truth, ie only 1 card per reflection.
The logic I'd use is something like this:

- when the viewer has spotlight open & a remote user starts to drag a card, make sure the remote user is not dragging the target reflection. if they are & the other user wins the arbitration, close the spotlight. see https://github.com/ParabolInc/parabol/blob/8c68bf771a350b31c7948e6bd11c5c110e30c977/packages/client/mutations/StartDraggingReflectionMutation.ts#L71-L82
- when the remote user is dragging a card, check to see if that card exists inside the spotlight. 
- if it does, animate the card in front of the spotlight modal
- it it does not, check the `remoteDrag.targetId` 
- if the targetId is null or the targetId does not exist in the spotlight, animate the remote reflection behind the spotlight modal
- if the targetId exists in the spotlight, animate the remote reflection relative to the reflection.

In plain english, let's assume you have a spotlight modal open. if a remote user is dragging a card `A` & hovering over `B`, then you should see that I'm hovering over `B` if it's in your spotlight results. that way when the remote user drops it, it doesn't seem like magic just happened.  

> Should the results be groupable/ungroupable?

- yes, you can drag a result card onto the target card OR you can drag the target card onto a result card.
- like you say, if a remote user drags a result card `A` onto a card that doesn't exist in spotlight `B`, then the spotlight should remove group `A` from the results & add group `B` in its place. visually, it'll look like the card just grew a stack behind it.
- if a remote user drags a card that doesn't exist in spotlight `B` onto a card that _does_ exist in spotlight results `A`, then the user should see that card hovering & then eventually dropped onto `A` (the logic is the same as your question above)

File: packages/server/graphql/types/User.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tianrunhe) [reply packages/server/graphql/types/User.ts:443]: > If the viewer passes a reflectionId to the query that lives in a meeting that they're not a part of, the viewer's meeting id should be different.

How do we test this case in the code?

We deconstruct the `meetingId` from the reflection in line 430 and then use that to construct a `meetingMemberId` in line 431. If the user is **not** in that meeting, then line 434 should return a null object, which will cause line 442 blow up, yes? On the other hand, if the use is indeed in that meeting, the returned `viewerMeetingMember` object will always have the same `meetingId` as the one we deconstructed from line 430, so the check in line 443 is not helpful.
Tell me if I go completely nut and become blinded in this part of code...

File: packages/server/graphql/types/User.ts [hunk 11]
```diff
@@ -442,37 +443,38 @@ const User = new GraphQLObjectType<any, GQLContext>({
         if (viewerMeetingId !== meetingId) {
           return standardError(new Error('Not on team'), {userId})
         }
-        let currentThreshold: number | null = AUTO_GROUPING_THRESHOLD
-        const similarGroupIds = new Set<string>()
         const reflectionsCount = reflections.length
-        const maxGroupSize = Math.min(reflectionsCount, 10)
-        while (currentThreshold) {
+        const maxReductionPercent = 1
+        const maxResultGroupSize = 10
```
Comment (tianrunhe) [inline packages/server/graphql/types/User.ts:448]: Sorry I didn't make it clear... I was proposing to put those in constants like `AUTO_GROUPING_THRESHOLD`

File: packages/client/components/ReflectionGroup/RemoteReflection.tsx [hunk 12]
```diff
@@ -103,16 +106,41 @@ const getHeaderTransform = (ref: RefObject<HTMLDivElement>, topPadding = 18) =>
...
-  const {left, top, minTop} = getCoords(remoteDrag as any)
-  return {nextStyle: {transform: `translate(${left}px,${top}px)`}, minTop}
+  const {left, top, minTop} = getCoords(remoteDrag as any, showAboveSpotlight)
+  const zIndex = showAboveSpotlight ? ZIndex.REFLECTION_IN_FLIGHT_SPOTLIGHT : style.zIndex
+  return {nextStyle: {transform: `translate(${left}px,${top}px)`, zIndex}, minTop}
+}
+
+const getTarget = (meetingId: string, targetId: string, atmosphere: Atmosphere) => {
+  let isTargetInSpotlight = false
+  commitLocalUpdate(atmosphere, (store) => {
```
Comment (nickoferrall) [inline packages/client/components/ReflectionGroup/RemoteReflection.tsx:120]: We're not updating the store here - just grabbing a value. I prefer using the Relay store instead of adding another Root query as it saves a trip to the server, and getting the required `reflectionId` and `searchQuery` variables for a Root query can be ugly. 

In [ErrorBoundary](https://github.com/ParabolInc/parabol/blob/89f7d0cceae2c30d7e8757767418ddade4ae98c1/packages/client/components/ErrorBoundary.tsx), I see we use `atmosphere.getStore()`, cast the `store` to `any`, and can then access the records, but I think this is slightly easier to read

Comment (nickoferrall) [reply packages/client/components/ReflectionGroup/RemoteReflection.tsx:120]: Try useQueryLoader: https://relay.dev/docs/guided-tour/reusing-cached-data/fetch-policies/

File: packages/client/components/SpotlightModal.tsx [hunk 13]
```diff
@@ -168,9 +208,11 @@ const SpotlightModal = (props: Props) => {
             />
           </SearchItem>
         </SelectedReflectionSection>
-        <SimilarReflectionGroups>
-          {reflectionGroupsCount === 0 ? <SpotlightEmptyState /> : null}
-        </SimilarReflectionGroups>
+        <SimilarGroups>
+          <Suspense fallback={''}>
```
Comment (nickoferrall) [inline packages/client/components/SpotlightModal.tsx:212]: Suspense is required here as the query in `SpotlightGroup` can be loading. If we remove this, remotely group & ungroup results in the Spotlight, notice that the modal flashes

File: packages/client/hooks/useDraggableReflectionCard.tsx [hunk 14]
```diff
@@ -33,19 +33,37 @@ const useRemoteDrag = (
...
   const setPortal = useContext(PortalContext)
-  const {remoteDrag, isDropping} = reflection
+  const {remoteDrag, isDropping, reflectionGroupId} = reflection
   const setRemoteCard = (isClose: boolean, timeRemaining: number, lastTop?: number) => {
     if (!drag.ref || timeRemaining <= 0) return
     const beforeFrame = Date.now()
     const bbox = drag.ref.getBoundingClientRect()
     if (bbox.top !== lastTop) {
+      const targetId = remoteDrag?.targetId
+      const isTargetInSpotlight = !!document.querySelector(
```
Comment (nickoferrall) [inline packages/client/hooks/useDraggableReflectionCard.tsx:43]: @mattkrick I looked into using a ref instead of the `querySelector`, but I think we need to have the logic within `setRemoteCard` as it's recursively updating with every animation frame. In `DraggableReflectionCard.tsx`, we could use the `targetId` to grab the target element and store it in `drag`,  but it won't update as quickly as we need

File: packages/server/graphql/types/User.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/types/User.ts]: True @tianrunhe! I've updated the check so that the error is thrown if `viewerMeetingMember` is null. I tested this by changing `reflectionId` to an id that belongs in a meeting that the viewer is not a part of. I'll move this to Maintainer review unless there's you have any other feedback?

**Comment by nickoferrall** [comment]
FYI: Two issues that I've noticed: 1. After closing the Spotlight, the reflections on the kanban can be stacked incorrectly. 2. The remote reflection position is sometimes slightly off - it looks like it occasionally uses the kanban reflection position instead of the Spotlight. This PR has a fair amount of bloat already, so I'd prefer to press on and tackle these in a separate issue

File: packages/client/components/ReflectionGroup/DraggableReflectionCard.tsx [hunk 15]
```diff
@@ -77,11 +103,15 @@ const DraggableReflectionCard = (props: Props) => {
...
+  const {id: meetingId, teamId, localStage, spotlightReflection} = meeting
   const {isComplete, phaseType} = localStage
   const {isDropping, isEditing} = reflection
-  const [drag] = useState(makeDragState)
+  const isSpotlightOpen = !!spotlightReflection?.id
+  const isInSpotlight = isSpotlightOpen && !openSpotlight
+  const isBehindSpotlight = isSpotlightOpen && !isInSpotlight
   const staticReflectionCount = staticReflections?.length || 0
+  const [drag] = useState(makeDragState)
+  updateSpotlightDrag(drag, isBehindSpotlight, isInSpotlight, spotlightReflection?.id)
```
Comment (nickoferrall) [inline packages/client/components/ReflectionGroup/DraggableReflectionCard.tsx:114]: I tried passing in a few args to `makeDragState` and updating the drag object with Spotlight logic that way, but it would update too slowly and throw off the UI

File: packages/client/mutations/handlers/handleUpdateSpotlightResults.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    })
+  }
+  // ungrouping created a new group id which needs to be added to Spotlight
+  else if (reflectionsCount === 1) {
+    const sortOrders = similarReflectionGroups.map((group) => ({
+      sortOrder: group.getValue('sortOrder') as number
+    }))
+    const nextSortOrder = getNextSortOrder(sortOrders)
+    reflectionGroup.setValue(nextSortOrder, 'sortOrder')
+    addNodeToArray(reflectionGroup, viewer, 'similarReflectionGroups', 'sortOrder', {
```
Comment (nickoferrall) [inline packages/client/mutations/handlers/handleUpdateSpotlightResults.ts:48]: In the groups to source PR, I noticed that sometimes this can cause an infinite loop when ungrouping. If I removed `handleUpdateSpotlightResults` from the optimistic updater, the infinite loop would stop happening. Not sure if there's anything wrong with the way I'm using the `addNodeToArray` function? If not, there's probably other logic in later PRs causing this issue. It's not happening for me in this branch.

File: packages/client/hooks/useDraggableReflectionCard.tsx (hunk 14 above)
Comment (mattkrick) [reply packages/client/hooks/useDraggableReflectionCard.tsx:43]: thanks for digging in!
inside hooks it should be AOK, if there's a better way to do it inside the a functional component let's try that. trying to access the DOM while computing what the vdom should look like might get dicey

possible ideas:

- `useState(() => document.querySelector(...))`
- `useRef` + `useEffect(() => { if (blah blah) document.querySelector}, [])`

File: packages/client/components/ReflectionCard/ReflectionCard.tsx (hunk 7 above)
Comment (mattkrick) [reply packages/client/components/ReflectionCard/ReflectionCard.tsx:233]: i think it's worth showing off, but we can save that for a later time

File: packages/client/shared/gqlIds/MeetingMemberId.ts [hunk 17]
```diff
@@ -1,7 +1,6 @@
-
 const MeetingMemberId = {
-  join: (meetingId: string, userId: string) => `${userId}::${meetingId}`,
-  split: (id: string) => {
+  join: ({meetingId, userId}: {meetingId: string; userId: string}) => `${userId}::${meetingId}`,
```
Comment (mattkrick) [inline packages/client/shared/gqlIds/MeetingMemberId.ts:2]: are you opposed to keeping the same API as before with 2 params? Typescript will still tell us which is which & then `MeetingMemberId.join` has the same signature as all the other gqlIds

File: packages/server/graphql/types/User.ts [hunk 18]
```diff
@@ -442,19 +442,55 @@ const User: GraphQLObjectType<any, GQLContext> = new GraphQLObjectType<any, GQLC
...
+          for (const groupedReflection of groupedReflectionsRes) {
+            const {reflectionGroupId, oldReflectionGroupId} = groupedReflection
+            if (
+              reflectionGroupId === spotlightGroup.reflectionGroupId &&
+              oldReflectionGroupId !== spotlightGroup.oldReflectionGroupId
+            ) {
+              spotlightResultGroupIds.add(oldReflectionGroupId)
+            }
+            currentThresh = nextThresh
+            if (spotlightResultGroupIds.size >= spotlightResultGroupSize) break
```
Comment (mattkrick) [inline packages/server/graphql/types/User.ts:486]: don't we want to break before the maximum is reached?
- save the results as R1
- compute next results R2
- if size R2 > max, return R1
- else R1 = R2

**Review by mattkrick — CHANGES_REQUESTED** [review]
nice work! couple housekeeping things, i think only having 1 droppable attribute should greatly clean up a lot of the complexity here.

File: packages/client/shared/gqlIds/MeetingMemberId.ts (hunk 17 above)
Comment (nickoferrall) [reply packages/client/shared/gqlIds/MeetingMemberId.ts]: I kinda like passing in an object when two args are of the same type as it prevents us from mixing up the order. But I'm happy to keep things consistent and revert that change 👌
